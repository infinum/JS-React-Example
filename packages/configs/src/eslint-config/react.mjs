import { fixupPluginRules } from '@eslint/compat';
import pluginReact from 'eslint-plugin-react';
import pluginReactHooks from 'eslint-plugin-react-hooks';

export default [
	{
		files: ['**/*.tsx', '**/*.jsx'],
		settings: {
			react: {
				version: 'detect',
			},
		},
		plugins: {
			react: pluginReact,
			'react-hooks': fixupPluginRules(pluginReactHooks),
		},
		rules: {
			...pluginReact.configs['jsx-runtime'].rules,
			...pluginReactHooks.configs.recommended.rules,
		},
	},
];
