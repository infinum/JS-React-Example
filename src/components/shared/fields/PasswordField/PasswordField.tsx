import {
	FormControl,
	FormErrorMessage,
	FormLabel,
	forwardRef,
	IconButton,
	Input,
	InputGroup,
	InputProps,
	InputRightElement,
	useDisclosure,
	useMergeRefs,
} from '@chakra-ui/react';
import { ErrorMessage } from '@hookform/error-message';
import { useRef } from 'react';
import { FieldErrors } from 'react-hook-form';
import { HiEye, HiEyeOff } from 'react-icons/hi';

export interface IPasswordField extends InputProps {
	label: string;
	errors?: FieldErrors;
	name: string;
}

export const PasswordField = forwardRef<IPasswordField, 'input'>(({ label, errors, name, id, ...rest }, ref) => {
	const { isOpen, onToggle } = useDisclosure();
	const inputRef = useRef<HTMLInputElement>(null);

	const mergeRef = useMergeRefs(inputRef, ref);
	const onClickReveal = () => {
		onToggle();

		if (inputRef.current) {
			inputRef.current.focus({ preventScroll: true });
		}
	};

	return (
		<FormControl id={id} isInvalid={Boolean(errors?.[name])}>
			<FormLabel>{label}</FormLabel>
			<InputGroup>
				<Input
					ref={mergeRef}
					autoComplete="current-password"
					name={name}
					required
					type={isOpen ? 'text' : 'password'}
					{...rest}
				/>
				<InputRightElement>
					<IconButton
						aria-label={isOpen ? 'Mask password' : 'Reveal password'}
						icon={isOpen ? <HiEyeOff /> : <HiEye />}
						onClick={onClickReveal}
						variant="link"
					/>
				</InputRightElement>
			</InputGroup>
			<FormErrorMessage color="red">
				<ErrorMessage errors={errors} name={name} />
			</FormErrorMessage>
		</FormControl>
	);
});
