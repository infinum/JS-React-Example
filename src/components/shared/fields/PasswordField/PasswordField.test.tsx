import { PasswordField } from '@/components/shared/fields/PasswordField/PasswordField';
import { FC } from 'react';
import { render, screen } from 'test-utils';
import userEvent from '@testing-library/user-event';

const label = 'Test Label';
const name = 'Test Name';

const TestPasswordField: FC = () => {
	return <PasswordField label={label} name={name} />;
};

describe('PasswordField', () => {
	it('should matches snapshot', () => {
		const { asFragment } = render(<TestPasswordField />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should render correct label', () => {
		render(<TestPasswordField />);

		expect(screen.getByText(label)).toBeInTheDocument();
	});

	it('should render correct name', () => {
		render(<TestPasswordField />);

		const passwordField = screen.getByLabelText(label);

		expect(passwordField).toHaveAttribute('name', name);
	});

	it('should toggle password visibility', async () => {
		const user = userEvent.setup();

		render(<TestPasswordField />);

		const toggleVisibilityButton = screen.getByRole('button', { name: /Reveal password/i });
		const passwordField = screen.getByLabelText(label);

		expect(toggleVisibilityButton).toBeInTheDocument();
		expect(passwordField).toHaveAttribute('type', 'password');

		await user.click(toggleVisibilityButton);

		expect(toggleVisibilityButton).toHaveAccessibleName('Mask password');
		expect(passwordField).toHaveAttribute('type', 'text');
	});
});
