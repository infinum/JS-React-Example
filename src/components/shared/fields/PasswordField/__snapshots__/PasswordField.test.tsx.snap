// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordField should matches snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  width: 100%;
  position: relative;
}

.emotion-1 {
  display: block;
  text-align: start;
  font-size: var(--chakra-fontSizes-md);
  -webkit-margin-end: var(--chakra-space-3);
  margin-inline-end: var(--chakra-space-3);
  margin-bottom: var(--chakra-space-2);
  font-weight: var(--chakra-fontWeights-medium);
  transition-property: var(--chakra-transition-property-common);
  transition-duration: var(--chakra-transition-duration-normal);
  opacity: 1;
}

.emotion-1:disabled,
.emotion-1[disabled],
.emotion-1[aria-disabled=true],
.emotion-1[data-disabled] {
  opacity: 0.4;
}

.emotion-2 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  isolation: isolate;
  --input-font-size: var(--chakra-fontSizes-md);
  --input-padding: var(--chakra-space-4);
  --input-border-radius: var(--chakra-radii-md);
  --input-height: var(--chakra-sizes-10);
}

.emotion-3 {
  width: 100%;
  height: var(--input-height);
  font-size: var(--input-font-size);
  -webkit-padding-start: var(--input-padding);
  padding-inline-start: var(--input-padding);
  -webkit-padding-end: var(--input-height);
  padding-inline-end: var(--input-height);
  border-radius: var(--input-border-radius);
  min-width: 0px;
  outline: 2px solid transparent;
  outline-offset: 2px;
  position: relative;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  transition-property: var(--chakra-transition-property-common);
  transition-duration: var(--chakra-transition-duration-normal);
  --input-font-size: var(--chakra-fontSizes-md);
  --input-padding: var(--chakra-space-4);
  --input-border-radius: var(--chakra-radii-md);
  --input-height: var(--chakra-sizes-10);
  border: 1px solid;
  border-color: inherit;
  background: inherit;
}

.emotion-3:disabled,
.emotion-3[disabled],
.emotion-3[aria-disabled=true],
.emotion-3[data-disabled] {
  opacity: 0.4;
  cursor: not-allowed;
}

.emotion-3:hover,
.emotion-3[data-hover] {
  border-color: var(--chakra-colors-gray-300);
}

.emotion-3[aria-readonly=true],
.emotion-3[readonly],
.emotion-3[data-readonly] {
  box-shadow: var(--chakra-shadows-none)!important;
  -webkit-user-select: all;
  -moz-user-select: all;
  -ms-user-select: all;
  user-select: all;
}

.emotion-3[aria-invalid=true],
.emotion-3[data-invalid] {
  border-color: #E53E3E;
  box-shadow: 0 0 0 1px #E53E3E;
}

.emotion-3:focus-visible,
.emotion-3[data-focus-visible] {
  z-index: 1;
  border-color: #3182ce;
  box-shadow: 0 0 0 1px #3182ce;
}

.emotion-4 {
  right: 0;
  width: var(--input-height);
  height: var(--input-height);
  font-size: var(--input-font-size);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: absolute;
  top: 0px;
  z-index: 2;
}

.emotion-5 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: baseline;
  outline: 2px solid transparent;
  outline-offset: 2px;
  line-height: var(--chakra-lineHeights-normal);
  border-radius: var(--chakra-radii-md);
  font-weight: var(--chakra-fontWeights-semibold);
  transition-property: var(--chakra-transition-property-common);
  transition-duration: var(--chakra-transition-duration-normal);
  height: auto;
  min-width: var(--chakra-sizes-10);
  font-size: var(--chakra-fontSizes-md);
  -webkit-padding-start: var(--chakra-space-4);
  padding-inline-start: var(--chakra-space-4);
  -webkit-padding-end: var(--chakra-space-4);
  padding-inline-end: var(--chakra-space-4);
  color: var(--chakra-colors-gray-500);
  padding: 0px;
}

.emotion-5:focus-visible,
.emotion-5[data-focus-visible] {
  box-shadow: var(--chakra-shadows-outline);
}

.emotion-5:disabled,
.emotion-5[disabled],
.emotion-5[aria-disabled=true],
.emotion-5[data-disabled] {
  opacity: 0.4;
  cursor: not-allowed;
  box-shadow: var(--chakra-shadows-none);
}

.emotion-5:hover,
.emotion-5[data-hover] {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-5:hover:disabled,
.emotion-5[data-hover]:disabled,
.emotion-5:hover[disabled],
.emotion-5[data-hover][disabled],
.emotion-5:hover[aria-disabled=true],
.emotion-5[data-hover][aria-disabled=true],
.emotion-5:hover[data-disabled],
.emotion-5[data-hover][data-disabled] {
  background: initial;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-5:active,
.emotion-5[data-active] {
  color: var(--chakra-colors-gray-700);
}

<div
    class="chakra-form-control emotion-0"
    role="group"
  >
    <label
      class="chakra-form__label emotion-1"
      for="field-:r1:"
      id="field-:r1:-label"
    >
      Test Label
    </label>
    <div
      class="chakra-input__group emotion-2"
      data-group="true"
    >
      <input
        aria-required="true"
        autocomplete="current-password"
        class="chakra-input emotion-3"
        id="field-:r1:"
        name="Test Name"
        required=""
        type="password"
      />
      <div
        class="chakra-input__right-element emotion-4"
      >
        <button
          aria-label="Reveal password"
          class="chakra-button emotion-5"
          type="button"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 20 20"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 12a2 2 0 100-4 2 2 0 000 4z"
            />
            <path
              clip-rule="evenodd"
              d="M.458 10C1.732 5.943 5.522 3 10 3s8.268 2.943 9.542 7c-1.274 4.057-5.064 7-9.542 7S1.732 14.057.458 10zM14 10a4 4 0 11-8 0 4 4 0 018 0z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
